### ISSUE TITLE
False positive for `enum_clike_unportable_variant` for values that are variable based on the width of a `usize`

### ISSUE THREAD

**Issue by DjLogozzo** [rust-lang/rust-clippy#8043]
Lint name:
enum_clike_unportable_variant

I tried this code:

```rust
#[repr(usize)]
pub enum Foo {
    Bar = 0,
    Baz = HIGH_BIT | 1,
}

const HIGH_BIT: usize = 1usize.rotate_right(1);
```

I expected to see this happen:
Since the value of `HIGH_BIT` is variable dependent on the size of `usize`, it should never overflow a `usize`.
That code is portable to 32-bit architectures because when compiled for a 32-bit arch, the value of `HIGH_BIT` would change to fit within a 32-width `usize`

Instead, this happened: 
```
error: C-like enum variant discriminant is not portable to 32-bit targets
 --> src/lib.rs:4:5
  |
4 |     Baz = HIGH_BIT | 1,
  |     ^^^^^^^^^^^^^^^^^^
  |
  = note: `#[deny(clippy::enum_clike_unportable_variant)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#enum_clike_unportable_variant

```

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.58.0-nightly (a77da2d45 2021-11-19)
binary: rustc
commit-hash: a77da2d454e6caa227a85b16410b95f93495e7e0
commit-date: 2021-11-19
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0
```

**Comment by DjLogozzo** [comment]
Possibly related issue: #1134
The issue I'm having looks like its related to this issue, where the value computed for the enum value is not taking into account that the value might be different on other architectures

**Comment by DjLogozzo** [comment]
When running clippy with a 32-bit target triple, it no-longer errors out

64-bit Architecture
```
deejay@laptop ~/src/helloworld $ cargo clippy --target x86_64-unknown-linux-gnu
    Checking helloworld v0.1.0 (/home/deejay/src/helloworld)
error: C-like enum variant discriminant is not portable to 32-bit targets
 --> src/main.rs:6:5
  |
6 |     Baz = HIGH_BIT | 1,
  |     ^^^^^^^^^^^^^^^^^^
  |
  = note: `#[deny(clippy::enum_clike_unportable_variant)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#enum_clike_unportable_variant

error: could not compile `helloworld` due to previous error
```
32-bit Architecture
```
deejay@laptop ~/src/helloworld $ cargo clippy --target i686-unknown-linux-gnu
    Checking helloworld v0.1.0 (/home/deejay/src/helloworld)
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
```

**Comment by dimpolo** [comment]
Minimized:
```rust
#[repr(usize)]
enum Foo {
    A = usize::MAX,
}
```

**Comment by tobz** [comment]
Just a fellow traveler stopping by to report this false positive is still present, nearly three years later. 😓

**Comment by Nyde42** [comment]
Yet another fellow time traveler - same error is still present on Rust 1.84.1 with the same example (except using `isize::MIN` instead of `usize::MAX`)
